### DISCUSSION TITLE
Imagining "PrairieLearn Live": An audience response system

### DISCUSSION THREAD

**Discussion by firasm** [PrairieLearn/PrairieLearn#6327] [Ideas]
One thing that would be **amazing** if PrairieLearn could do, is act as an "Audience Response System" similar to other commercial products that exist like [Sli.do](https://sli.do), [iClicker](http://iclicker.com), [PollEV](http://Pollev.com), etc... (there are a dime a dozen of these). 

We were encouraged to create a discussion about this to start brainstorming what this may look like.

## Use-case

- Active learning in classes to do "clicker" and "worksheet" questions in class.

### Required

- Live Statistics on what students choose
- Instructor-controls when questions get released to students, when/if results get released to students
- Use existing PL elements with a "flag" in the `infoAssessment.json` file to "enable" live class version and turn off randomization
- Ability to easily "run the question again" as a different version (After an initial vote, keeping both initial and secondary votes)

### Optional Nice-to-haves

- Instructor "Present" view (what students should see when instructor shares their screen) and "Admin" view (what instructor sees on their side)
- Add a Timer so students know how much time they have to complete the question
- Accounting for Randomization of questions (MVP is to do it without considering randomization of MCQs etc...)
- Student UI to see the vote results (Usually an instructor is sharing their screen so this isn't really needed)
- Summary Report of Clicker questions at the end of the class

### Option 1: Embed an iFrame in a PL question:

![prairielearn_slido_demo](https://user-images.githubusercontent.com/2507459/190237284-f8391413-e3ad-4f2c-a5fb-2ec778387588.gif)

**Pros**
- Easy to implement
- Works out of the box
- No headache for students, looks like a "PL assignment/homework/worksheet"
- Flexible and can use several tools
- Instructor-paced
- 
**Cons**
- Tracking students has to be done separately (in a different app)
- No way to disable the "save and grade" button so somewhat confusing where to click sometimes

### Option 2: Use API requests externally

Details coming....See issue #5644

**Comment by jonatanschroeder** [comment]
I can see this as a separate "mode" (instead of Exam or Public mode), this may be one way to handle it.

**Reply by ffund** [reply]
Would the idea here be to have an existing question (in "native" PL format) be answer-able by a clicker or phone UI? That would be really cool - to be able to

* ask a homework question in class and poll students live
* discuss the question and answer in class
* students do the same question individually in a later HW or Exam assessment

**Reply by jonatanschroeder** [reply]
The PL web interface is already relatively responsive in a mobile browser UI, so I don't think a lot of stuff needs to be done in terms of student interface. This is mainly in regards to giving instructor control as to (a) when a question is enabled, (b) when the answer is given, and (c) what the pace of multiple questions is. Then instructors should be able to see stats on individual answers live, including which answers were more common, particularly for MC, dropdown, checkbox, matching questions. But your list of items is exactly what we have in mind.

**Comment by echuber2** [comment]
Are there any risks for clickjacking if we set a precedent of embedding iframes of external domains? (Maybe someone is already doing this, I don't know.)

If it's called PrairieLearn Live we _definitely_ need to add Achievements.
